rust_src/src/eval.rs

```diff
@@ -1,91 +1,320 @@
-macro_rules! error {
-    ($str:expr) => {{
-        let strobj = unsafe {
-            ::remacs_sys::make_string($str.as_ptr() as *const ::libc::c_char,
-                                      $str.len() as ::libc::ptrdiff_t)
-        };
-        xsignal!(::remacs_sys::Qerror, $crate::lisp::LispObject::from_raw(strobj));
-    }};
-    ($fmtstr:expr, $($arg:expr),*) => {{
-        let formatted = format!($fmtstr, $($arg),*);
-        let strobj = unsafe {
-            ::remacs_sys::make_string(formatted.as_ptr() as *const ::libc::c_char,
-                                      formatted.len() as ::libc::ptrdiff_t)
-        };
-        xsignal!(::remacs_sys::Qerror, $crate::lisp::LispObject::from_raw(strobj));
-    }};
+    quoted
 }
+def_lisp_sym!(Qfunction, "function");
 
-/// Macro to format a "wrong argument type" error message.
-macro_rules! wrong_type {
-    ($pred:expr, $arg:expr) => {
-        xsignal!(::remacs_sys::Qwrong_type_argument, LispObject::from_raw($pred), $arg);
-    };
+// Make SYMBOL lexically scoped.
+/// Internal function
+#[lisp_fn(name = "internal-make-var-non-special")]
+pub fn make_var_non_special(symbol: LispSymbolRef) -> bool {
+    symbol.set_declared_special(false);
+    true
 }
 
-macro_rules! args_out_of_range {
-    ($($tt:tt)+) => { xsignal!(::remacs_sys::Qargs_out_of_range, $($tt)+); };
+/// Return non-nil if SYMBOL's global binding has been declared special.
+/// A special variable is one that will be bound dynamically, even in a
+/// context where binding is lexical by default.
+#[lisp_fn]
+pub fn special_variable_p(symbol: LispSymbolRef) -> bool {
+    symbol.get_declared_special()
 }
 
-macro_rules! list {
-    ($arg:expr, $($tt:tt)+) => { $crate::lisp::LispObject::cons($arg, list!($($tt)+)) };
-    ($arg:expr) => { $crate::lisp::LispObject::cons($arg, list!()) };
-    () => { $crate::lisp::LispObject::constant_nil() };
+/// Define SYMBOL as a constant variable.
+/// This declares that neither programs nor users should ever change the
+/// value.  This constancy is not actually enforced by Emacs Lisp, but
+/// SYMBOL is marked as a special variable so that it is never lexically
+/// bound.
+///
+/// The `defconst' form always sets the value of SYMBOL to the result of
+/// evalling INITVALUE.  If SYMBOL is buffer-local, its default value is
+/// what is set; buffer-local values are not affected.  If SYMBOL has a
+/// local binding, then this form sets the local binding's value.
+/// However, you should normally not make local bindings for variables
+/// defined with this form.
+///
+/// The optional DOCSTRING specifies the variable's documentation string.
+/// usage: (defconst SYMBOL INITVALUE [DOCSTRING])
+#[lisp_fn(min = "2", unevalled = "true")]
+pub fn defconst(args: LispObject) -> LispSymbolRef {
+    let (sym, tail) = args.as_cons_or_error().as_tuple();
+
+    let mut docstring = LispObject::constant_nil();
+    if cdr(tail).is_not_nil() {
+        if cdr(cdr(tail)).is_not_nil() {
```

Comment:
Unlike the C, I'm sure this is correct as `cdr` checks the type of its argument :+1: